### PR TITLE
Added QuantileIVLinearEstimator

### PR DESCRIPTION
**Pull request by hanzak**
Added QuantileIVLinearEstimator class to perform the linear inference strategy

### CODE REVIEW

File: ml_mr/estimation/quantile_iv.py [hunk 1]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+    representation of the exposure learned by the outcome neural network.
+    """
+    def __init__(
+        self,
+        dataset: IVDataset,
+        parent: QuantileIVEstimator
+    ):
+        self.parent = parent
+        self.exposure_network = parent.exposure_network
+        self.outcome_network = parent.outcome_network
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:308]: Unneeded. I think it will be less error prone to index from the parent directly.

File: ml_mr/estimation/quantile_iv.py [hunk 2]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+        self.covars = parent.covars
+        self.pca = self.init_PCA()
+        self.dataset = dataset
+        self.h = self.linear_inference(dataset)
+
+    def iv_reg_function(
+        self,
+        x: torch.Tensor,
+        covars: Optional[torch.Tensor] = None,
+        alpha: float = 0.1
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:318]: I will transition the default to 0.05 everywhere when I have time. I suggest you apply it here already.

```suggestion
        alpha: float = 0.05
```

File: ml_mr/estimation/quantile_iv.py [hunk 3]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+        dataset: IVDataset,
+        parent: QuantileIVEstimator
+    ):
+        self.parent = parent
+        self.exposure_network = parent.exposure_network
+        self.outcome_network = parent.outcome_network
+        self.covars = parent.covars
+        self.pca = self.init_PCA()
+        self.dataset = dataset
+        self.h = self.linear_inference(dataset)
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:312]: I suggest addind a type annotation here as a form of documentation. It seems like h is a callable that returns the mean and se of h hat, but it would be good to make it explicit.

File: ml_mr/estimation/quantile_iv.py [hunk 4]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+        x: torch.Tensor,
+        covars: Optional[torch.Tensor] = None,
+        alpha: float = 0.1
+    ):
+        ys, se_ys = self.h(x, covars)
+        lower_ci = ys + stats.norm.ppf(alpha/2) * se_ys
+        upper_ci = ys + stats.norm.ppf(1-alpha/2) * se_ys
+
+        return torch.hstack([lower_ci, ys, upper_ci]).view(
+            len(ys), 1, 3
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:325]: FYI you can use -1 to "fill in" dimensions:

```suggestion
            -1, 1, 3
```

File: ml_mr/estimation/quantile_iv.py [hunk 5]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+            x: torch.Tensor,
+            covars: Optional[torch.Tensor] = None,
+            low_memory: bool = True,
+            alpha: float = 0.1
+    ) -> torch.Tensor:
+        return super().avg_iv_reg_function(x, covars, low_memory, alpha)
+
+    def init_PCA(self, n_components: float = 0.999, random_state: int = 42):
+        pca = PCA(n_components=n_components, random_state=random_state)
+        return pca
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:339]: Can probably just initialize directly in __init__

File: ml_mr/estimation/quantile_iv.py [hunk 6]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+
+    def read_ivdataset(self, dataset: IVDataset) -> Tuple:
+        dl = FullBatchDataLoader(dataset)
+        x, y, ivs, covars = next(iter(dl))
+        return x, y, ivs, covars
+
+    def linear_inference(
+        self,
+        dataset: IVDataset
+    ) -> Callable[[torch.Tensor, Optional[torch.Tensor]], torch.Tensor]:
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:349]: I think the annotation is wrong. The return type is a callable that returns a Tuple

File: ml_mr/estimation/quantile_iv.py [hunk 7]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+
+            if eta_bar is None:
+                eta_bar = current_representation
+            else:
+                eta_bar += current_representation
+
+        assert eta_bar is not None
+
+        # Fit PCA
+        self.pca.fit(eta_bar.detach().numpy())
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:372]: Use detach + numpy only once.

File: ml_mr/estimation/quantile_iv.py [hunk 8]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+        # Apply PCA to the representation of the exposure x
+        H_pca = torch.from_numpy(
+            self.pca.transform(H.detach().numpy())
+        )
+
+        # Add intercept column
+        H_pca = torch.hstack((torch.ones((n, 1)), H_pca))
+
+        # We now compute the IV regression
+        beta_IV_hat = torch.linalg.inv(
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:392]: Use torch.linalg.lstsq to avoid inversion.

File: ml_mr/estimation/quantile_iv.py [hunk 9]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+            @ mat_etabar_H_inv
+        )
+
+        # Estimated h_hat function
+        def iv_reg(x, covars):
+            eta = self.outcome_network.get_representation(
+                _cat(x, covars)
+            )
+            eta_pca = torch.from_numpy(
+                self.pca.transform(eta.detach().numpy())
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:413]: Eventually, if we want to make the implementation cleaner, we could use torch.svd instead of deferring (and depending on) scikit-learn for the PCA.

File: ml_mr/estimation/quantile_iv.py [hunk 10]
```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
...
+            )
+            # If we have covariables, we marginalize over the covariates
+            if covars is not None:
+                eta_pca = torch.vstack([
+                    tens.mean(dim=0, keepdim=True)
+                    for tens in torch.split(eta_pca, covars.shape[0])
+                ])
```
Comment (legaultmarc) [inline ml_mr/estimation/quantile_iv.py:420]: Are you sure this is correct?
